### ISSUE TITLE
Add Mechanistic Topic Models (MTM): topics over sparse-autoencoder features (Zheng et al. 2025)

### ISSUE THREAD

**Issue by nealcaren** [nealcaren/topica#575]
## Model to add

**Mechanistic Topic Models (MTMs)** — "Model Directions, Not Words: Mechanistic Topic Models Using Sparse Autoencoders", Zheng, Beltran-Velez, Karlekar, Shi, Nazaret, Mallik, Feder & Blei (2025; arXiv:2507.23220, TACL forthcoming).
- Paper: https://arxiv.org/abs/2507.23220
- Reference code: https://github.com/blei-lab/mechanistic-topic-models (MIT license, Python/Poetry)

## What it is

Instead of modeling topics as distributions over **words** (bag-of-words, as in LDA/CTM/STM), MTM models topics as distributions over **interpretable features from a sparse autoencoder (SAE)** trained on an LLM's activations (the reference uses Gemma-2-9b SAE features). A document is represented by its SAE feature activations; topics are directions in that feature space. This yields:
- richer, more abstract topic descriptions (feature concepts, not just word lists), and
- **controllable generation via topic steering vectors** — the distinctive capability over classical/neural topic models.

Reference (MIT): consumes **pretrained** SAE features (it does not train the SAE); `scripts/train.py` / `scripts/evaluate.py` over config files; HuggingFace/OpenAI/DeepSeek APIs for feature extraction and LLM-based eval; evaluated on 8 datasets (20 Newsgroups, AG News, Yelp, Bills, Wiki, GoEmotions, PoemSum, WritingPrompts).

## How it fits topica

This is a natural fit for topica's **"you bring the features" pattern** — the same split we already use for the embedding-cluster and neural models (`Top2Vec`/`BERTopic`/`ETM`/`Scholar` take a precomputed embedding matrix; `topica.llm_embed` builds one). Concretely:

- **In scope for the Rust core:** the topic model *over a supplied document × SAE-feature matrix* (sparse, non-negative activations) — fit topics as feature distributions and documents as topic mixtures, with topica's numpy-native API and the shared contract (coherence, labeling, effects, save/load).
- **Out of the core (external, like `llm_embed`):** the SAE feature-extraction pipeline (LLM activations → pretrained SAE → feature matrix). Heavy and model-specific (Gemma-2-9b + a pretrained SAE); should stay a documented user-supplied input or an optional Python helper, not a Rust dependency. This keeps the core dependency-light (no PyTorch/transformers in Rust), consistent with how the VAE-family and embedding models are structured.
- **Steering vectors / controllable generation:** likely a follow-up PR — the first port can land the feature-space topic model + feature-based topic descriptions; generation-steering is a separable capability.

## Approach (following the add-topic-model workflow)

1. Read the paper + reference `src/mechanistic_topic_models/` to pin the exact generative model and inference (is it LDA-over-features, a logistic-normal/VAE over features, or a direct decomposition?), and build a gold-standard result set from the reference on a shared dataset (e.g. 20NG) with a fixed, committed SAE feature matrix.
2. Implement the Rust core (topics over features) under topica's conventions; Python binding takes the feature matrix (+ optional feature vocabulary/labels) the way `Top2Vec.fit(docs, doc_embeddings, ...)` does.
3. Validate with the independent-benchmark + author-emulation review agents; ship experimental-tier first (`topica.enable_experimental()`), graduating once parity is demonstrated.
4. Optional PR2: topic steering / controllable generation; optional helper for SAE feature extraction.

## Notes

- Ship gated in the **experimental tier** until paper+parity validation is done (per the experimental-model convention).
- The reference is MIT-licensed; the SAE feature files it ships are for Gemma-2-9b.
- Bib/paper citation for this model would need the PDF fetched under the project's citation rule (cite by Better BibTeX key + only sources we hold) — not required to open this issue, just flagging for whenever it lands in the roster docs.

Filed at Neal's request.

**Comment by nealcaren** [comment]
That makes sense. I would incorporate this into `topica` as a **new feature-representation layer plus three model wrappers**, rather than treating the SAE itself as another topic model.

The key design distinction is:

[
\text{SAE featurization}
\quad\neq\quad
\text{topic-model inference}
]

The paper first turns documents into a sparse document-by-feature count matrix and then applies mLDA, mETM, or mBERTopic.  That separation fits `topica` particularly well because the package already places many inference families behind a common `topic_word`/`doc_topic` interface. ([[GitHub](https://github.com/nealcaren/topica)][1])

## Recommended architecture

I would add an object resembling:

```python
features = topica.SAEFeatures.from_texts(
    texts,
    model="google/gemma-2-2b",
    sae="gemma-scope-2b-pt-res-canonical",
    layer=12,
    threshold="p80",
)
```

It would contain:

```python
features.counts          # documents × SAE features, sparse
features.activations     # optional aggregate activations
features.directions      # SAE decoder matrix, W × H
features.descriptions    # one textual label per feature
features.thresholds      # q_i
features.token_counts    # N_tok for each document
features.metadata        # model, layer, SAE release, filtering
```

Then the models would be:

```python
model = topica.MechanisticLDA(num_topics=20)
model.fit(features)
```

```python
model = topica.MechanisticETM(num_topics=20)
model.fit(features)
```

```python
model = topica.MechanisticBERTopic(num_topics=20)
model.fit(features)
```

I would avoid making users write:

```python
model.fit(texts, model_name=..., sae_name=..., layer=...)
```

That mixes a very costly, hardware-dependent featurization operation with comparatively ordinary topic-model fitting. Users should be able to generate the SAE representation once, save it, and fit many models and values of (K).

## The most important first step: an SAE feature corpus

`topica` currently centers its shared model API on `topic_word`, `doc_topic`, and `top_words()`, while its standard installation deliberately requires only NumPy and pandas and avoids PyTorch. ([[GitHub](https://github.com/nealcaren/topica)][1]) Therefore, the central addition should be a model-independent data structure rather than PyTorch code embedded in every estimator.

Something like:

```python
@dataclass
class FeatureCorpus:
    matrix: SparseMatrix
    feature_names: list[str]
    document_ids: np.ndarray
    document_lengths: np.ndarray | None = None
    feature_vectors: np.ndarray | None = None
    metadata: dict[str, Any] = field(default_factory=dict)
```

I would make it more general than `SAECorpus`. SAE features are one case of a broader useful abstraction:

* SAE feature counts
* concept counts
* entity counts
* dictionary-derived features
* image-codebook features
* multimodal features

Internally, though, `SAEFeatures` could be a specialized constructor or subclass that carries SAE-specific provenance.

## Phase 1: implement mLDA first

This should be relatively easy because the paper’s mLDA is mathematically just ordinary LDA with SAE features substituted for words:

[
\widetilde{\mathbf c}*d
\sim
\operatorname{Multinomial}
(\boldsymbol\theta_d\boldsymbol\beta, N*{\text{sae}})
]

The existing Rust SparseLDA implementation should not need a new sampler. The real work is allowing LDA to consume an already-constructed sparse count corpus whose vocabulary is a collection of feature IDs rather than words.

A plausible API:

```python
feature_corpus = topica.from_feature_matrix(
    counts,
    feature_names=descriptions,
    document_ids=document_ids,
)

model = topica.LDA(num_topics=20, seed=42)
model.fit(feature_corpus)
```

Then either:

```python
topica.MechanisticLDA = topica.LDA
```

with a semantic alias, or:

```python
class MechanisticLDA(LDA):
    ...
```

that validates SAE metadata and changes summaries from “top words” to “top features.”

I lean toward the thin subclass. It gives you a discoverable implementation of the published method and an appropriate summary without duplicating inference.

### Do not expand counts into repeated token IDs in Python

That would technically work, but it could greatly inflate the corpus because several SAE features may exceed threshold at a single token. A direct CSR/count-matrix constructor into the Rust corpus representation is the cleaner route.

## Phase 2: mBERTopic

This is also mostly composition of existing pieces.

The paper constructs each document representation as:

[
\widetilde{\mathbf e}_d
=======================

\frac{1}{N_{\text{tok},d}}
\sum_i \widetilde c_{d,i}\mathbf w_i
]

In matrix form:

[
E = \operatorname{diag}(N_{\text{tok}}^{-1})CW_{\text{out}}^\top
]

where:

* (C) is document-by-SAE-feature counts;
* (W_{\text{out}}^\top) is feature-by-hidden-dimension;
* (E) is document-by-hidden-dimension.

That can be a standalone method:

```python
embeddings = features.document_embeddings(
    weighting="count",
    normalize_by="tokens",
)
```

Then much of the current `BERTopic` pathway can be reused:

```python
model = topica.MechanisticBERTopic(num_topics=20)
model.fit(features)
```

Internally:

1. Reconstruct document embeddings using decoder directions.
2. Reduce and cluster using the existing BERTopic machinery.
3. Apply class-based TF–IDF to **SAE feature counts**, not original word counts.
4. Return topic-feature weights as the model’s principal topic representation.

`topica` already has a BERTopic implementation organized around embedding clustering and class-based TF–IDF, so this is a natural extension. ([[GitHub](https://github.com/nealcaren/topica)][1])

## Phase 3: mETM

This is the only genuinely new estimator.

The paper’s model is not just ordinary ETM with different embeddings. It learns topic activation vectors (\mathbf v_k) in the LLM’s hidden space and maps them into feature probabilities through the fixed SAE encoder:

[
\boldsymbol\beta_k
==================

\sigma(W_{\text{in}}\mathbf v_k+\mathbf b)
]

It also uses a binomial likelihood:

[
\widetilde c_{d,i}
\sim
\operatorname{Binomial}
\left(
[\boldsymbol\theta_d\boldsymbol\beta]*i,
N*{\text{tok},d}
\right)
]

That means mETM requires:

* the SAE encoder matrix, not merely decoder directions;
* a document token count;
* a binomial reconstruction objective;
* amortized variational inference;
* learned topic vectors in LLM activation space.

I would not force this through the existing ETM implementation. Share neural-network utilities where useful, but create a distinct Rust estimator and class.

```python
model = topica.MechanisticETM(
    num_topics=20,
    hidden_dim=128,
    seed=42,
)
model.fit(features)
```

The `FeatureCorpus` validation should reject input that lacks:

```python
features.encoder_weight
features.encoder_bias
features.token_counts
```

## SAE generation should be an optional extra

The base package currently advertises a small NumPy/pandas-only installation, with heavier capabilities placed in extras. ([[GitHub](https://github.com/nealcaren/topica)][1]) Keep that property.

I would use:

```toml
sae = [
    "torch>=2",
    "transformers>=...",
    "sae-lens>=..."
]
```

and perhaps:

```toml
sae-gemma = [...]
```

only if model-specific dependencies become awkward.

Usage:

```bash
pip install "topica[sae]"
```

The core package should still be able to **load, save, inspect, and model** precomputed SAE features without PyTorch. Only `SAEFeatures.from_texts()` would require the extra.

This matters because many users will generate the features on a GPU machine and analyze them elsewhere.

## Storage format

A fitted feature corpus could be fairly large. I would save:

```text
counts.npz
feature_directions.npy
encoder_weight.npy
encoder_bias.npy
thresholds.npy
token_counts.npy
feature_descriptions.json
documents.json
manifest.json
```

Potentially:

```python
features.save("corpus.sae")
features = topica.SAEFeatures.load("corpus.sae")
```

The manifest should record:

* source model and revision;
* tokenizer and revision;
* SAE release;
* layer and hook point;
* context length;
* treatment of special and padding tokens;
* threshold source and percentile;
* reference corpus used to estimate thresholds;
* feature filtering method;
* feature IDs retained;
* package versions.

Without that metadata, two SAE count matrices are not scientifically comparable.

## Feature descriptions versus “top words”

This is the largest API mismatch.

For mLDA and mETM:

```python
model.topic_word
```

is really:

```python
model.topic_feature
```

And:

```python
model.top_words()
```

would actually return feature descriptions.

I would expose both:

```python
model.topic_feature
model.top_features(n=10)
```

while maintaining:

```python
model.topic_word       # compatibility alias
model.top_words(n=10)  # compatibility alias
```

The summary should say:

```text
Topic 3
  labor organizing                     0.084
  collective bargaining disputes       0.071
  workplace protest and strikes        0.059
```

not pretend these are words.

You could also add a second lexical representation by collecting the words or spans associated with high feature activation:

```python
model.top_features()
model.top_terms()
```

The first reports SAE descriptions; the second translates topics back into corpus vocabulary. That would make the models easier to compare with ordinary topic models.

## Feature filtering should not be baked into inference

The paper uses an LLM to remove grammatical, spurious, and topic-irrelevant SAE features before fitting.  I would make that an explicit preprocessing method:

```python
features = features.filter(
    min_document_frequency=10,
    max_document_proportion=0.80,
)
```

and separately:

```python
features = topica.filter_sae_features(
    features,
    method="llm",
    backend=...,
)
```

Also provide reproducible non-LLM filters:

* minimum activation count;
* document frequency;
* maximum document frequency;
* feature-description keyword filters;
* variance;
* entropy across documents;
* exclusion of features marked as syntax or formatting;
* an explicit Boolean mask.

That would let you assess how much of the paper’s performance comes from the SAE and how much comes from LLM feature curation.

## A clean user workflow

The polished API might look like this:

```python
import topica

# GPU-dependent and potentially expensive.
features = topica.SAEFeatures.from_texts(
    df["text"],
    model="google/gemma-2-2b",
    sae="gemma-scope",
    layer=12,
    threshold_quantile=0.80,
    batch_size=8,
)

features = features.filter(
    min_document_frequency=10,
    max_document_proportion=0.75,
)

features.save("my_corpus.sae")

# Ordinary topica workflow from here onward.
model = topica.MechanisticLDA(num_topics=20, seed=42)
model.fit(features)

print(topica.summary(model))
print(model.top_features(10))

effect = topica.estimate_effect(
    model.doc_topic,
    df[["year", "party"]]
)
```

And for users with precomputed data:

```python
features = topica.from_feature_matrix(
    counts,
    feature_names=descriptions,
    feature_vectors=decoder_directions,
    token_counts=token_counts,
)
```

## Implementation order I would use

1. **Generic sparse `FeatureCorpus` and serialization.**
2. **Allow ordinary LDA to fit feature-count corpora.**
3. **Add `MechanisticLDA` as a thin semantic wrapper.**
4. **Add SAE-aware summaries and `top_features()`.**
5. **Add reconstructed SAE document embeddings.**
6. **Add `MechanisticBERTopic` by composing existing BERTopic parts.**
7. **Add optional SAE-Lens/Gemma Scope generation adapter.**
8. **Implement mETM last.**
9. **Only then add automatic LLM feature filtering and steering.**

The first six steps would already provide the two most straightforward variants from the paper and would not require changing `topica`’s lightweight core. I would classify all three as **experimental** initially, consistent with the repository’s current policy for implementations lacking published-reference parity. ([[GitHub](https://github.com/nealcaren/topica)][1])

The best immediate target is therefore not “add three new models.” It is: **make sparse semantic feature matrices first-class inputs in `topica`; then mLDA becomes almost free, mBERTopic becomes composition, and only mETM requires substantial new inference code.**

[1]: https://github.com/nealcaren/topica "GitHub - nealcaren/topica: Fast, all-purpose topic modeling for Python: 20+ models (LDA, STM, CTM, DMR, keyATM, BERTopic, ETM, ...) behind one NumPy-native interface on a parallel Rust core, with shared diagnostics and reproducible fits. · GitHub"

**Comment by nealcaren** [comment]
Related embedding-based topic-model triage (Neal's request, 2026-07-27): assessed two more papers and filed #580 (c-CluHTM hierarchical NMF over CluWords), #581 (CETopic TFIDF×IDF weighting — a BERTopic word-selection feature, not a model), #582 (Uniqueness + SHS HTM eval metrics). All three are distinct lineage from the SAE models (#575/#576).
